File: ShowCore/src/java/com/iflytek/cyber/iot/show/core/WifiInfoManager.java
package com.iflytek.cyber.iot.show.core;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkInfo;
import android.net.NetworkRequest;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class WifiInfoManager {

    private static final WifiInfoManager sManager = new WifiInfoManager();

    private List<NetworkStateListener> listeners = new ArrayList<>();
    private List<WifiRssiListener> rssiListeners = new ArrayList<>();

    private Handler uiHandler = new Handler(Looper.getMainLooper());

    public static WifiInfoManager getManager() {
        return sManager;
    }

    @Nullable
    public WifiInfo getWifiInfo(Context context) {
        final WifiManager manager = (WifiManager) context.getApplicationContext()
                .getSystemService(Context.WIFI_SERVICE);
        if (manager != null) {
            return manager.getConnectionInfo();
        } else {
            return null;
        }
    }

    public int getWifiSignalLevel(Context context) {
        final WifiInfo wifiInfo = getWifiInfo(context);
        if (wifiInfo != null) {
            return WifiManager.calculateSignalLevel(wifiInfo.getRssi(), 5);
        } else {
            return 0;
        }
    }

    public boolean isNetworkAvailable(Context context) {
        final ConnectivityManager manager = (ConnectivityManager) context.getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        if (manager != null) {
            final NetworkInfo networkInfo = manager.getActiveNetworkInfo();
            return networkInfo != null && networkInfo.isAvailable();
        } else {
            return false;
        }
    }

    public void registerNetworkCallback(Context context, @NonNull NetworkStateListener networkStateListener) {
        listeners.add(networkStateListener);
        final ConnectivityManager manager = (ConnectivityManager) context.getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        if (manager != null) {
            final NetworkRequest request = new NetworkRequest.Builder()
                    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build();
            manager.registerNetworkCallback(request, networkCallback);
        }
    }

    public void unregisterNetworkCallback(Context context) {
        final ConnectivityManager manager = (ConnectivityManager) context.getApplicationContext()
                .getSystemService(Context.CONNECTIVITY_SERVICE);
        if (manager != null) {
            manager.unregisterNetworkCallback(networkCallback);
        }
    }

    public void registerWifiRssiCallback(Context context, WifiRssiListener listener) {
        rssiListeners.add(listener);
        context.registerReceiver(receiver, new IntentFilter(WifiManager.RSSI_CHANGED_ACTION));
    }

    public void unregisterWifiRssiCallback(Context context) {
        if (receiver != null) {
            context.unregisterReceiver(receiver);
        }
        receiver = null;
    }

    private BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            for (WifiRssiListener listener : rssiListeners) {
                listener.onChange();
            }
        }
    };

    private ConnectivityManager.NetworkCallback networkCallback = new ConnectivityManager.NetworkCallback() {
        @Override
        public void onAvailable(Network network) {
            super.onAvailable(network);
            uiHandler.post(() -> {
                for (NetworkStateListener listener : listeners) {
                    listener.onAvailable(network);
                }
            });
        }

        @Override
        public void onLost(Network network) {
            super.onLost(network);
            uiHandler.post(() -> {
                for (NetworkStateListener listener : listeners) {
                    listener.onLost(network);
                }
            });
        }
    };

    public interface NetworkStateListener {
        void onAvailable(Network network);
        void onLost(Network network);
    }

    public interface WifiRssiListener {
        void onChange();
    }
}

File: app/src/main/java/com/iflytek/cyber/iot/show/core/utils/WifiInfoManager.kt
/*
 * Copyright (C) 2019 iFLYTEK CO.,LTD.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.iflytek.cyber.iot.show.core.utils

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import java.util.*

class WifiInfoManager private constructor() {

    private val listeners = ArrayList<NetworkStateListener>()
    private val rssiListeners = ArrayList<WifiRssiListener>()

    private val uiHandler = Handler(Looper.getMainLooper())

    private var receiver: BroadcastReceiver? = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            for (listener in rssiListeners) {
                listener.onChange()
            }
        }
    }

    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    init {
        if (Build.VERSION.SDK_INT >= 21)
            networkCallback = object : ConnectivityManager.NetworkCallback() {
                override fun onAvailable(network: Network) {
                    super.onAvailable(network)
                    uiHandler.post {
                        for (listener in listeners) {
                            listener.onAvailable(network)
                        }
                    }
                }

                override fun onLost(network: Network) {
                    super.onLost(network)
                    uiHandler.post {
                        for (listener in listeners) {
                            listener.onLost(network)
                        }
                    }
                }
            }
    }

    fun getWifiInfo(context: Context): WifiInfo? {
        val manager = context.applicationContext
                .getSystemService(Context.WIFI_SERVICE) as? WifiManager
        return manager?.connectionInfo
    }

    fun getWifiSignalLevel(context: Context): Int {
        val wifiInfo = getWifiInfo(context)
        return if (wifiInfo != null) {
            WifiManager.calculateSignalLevel(wifiInfo.rssi, 5)
        } else {
            0
        }
    }

    fun isNetworkAvailable(context: Context): Boolean {
        val manager = context.applicationContext
                .getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        return if (manager != null) {
            val networkInfo = manager.activeNetworkInfo
            networkInfo != null && networkInfo.isAvailable
        } else {
            false
        }
    }

    fun registerNetworkCallback(context: Context, networkStateListener: NetworkStateListener) {
        listeners.add(networkStateListener)
        val manager = context.applicationContext
                .getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        if (manager != null && Build.VERSION.SDK_INT >= 21) {
            val request = NetworkRequest.Builder()
                    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build()
            manager.registerNetworkCallback(request, networkCallback)
        }
    }

    fun unregisterNetworkCallback(context: Context) {
        val manager = context.applicationContext
                .getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
        if (manager != null && Build.VERSION.SDK_INT >= 21) {
            manager.unregisterNetworkCallback(networkCallback)
        }
    }

    fun registerWifiRssiCallback(context: Context, listener: WifiRssiListener) {
        rssiListeners.add(listener)
        context.registerReceiver(receiver, IntentFilter(WifiManager.RSSI_CHANGED_ACTION))
    }

    fun unregisterWifiRssiCallback(context: Context) {
        if (receiver != null) {
            context.unregisterReceiver(receiver)
        }
        receiver = null
    }

    interface NetworkStateListener {
        fun onAvailable(network: Network)

        fun onLost(network: Network)
    }

    interface WifiRssiListener {
        fun onChange()
    }

    companion object {

        val manager = WifiInfoManager()
    }

}
